extension-workspace-references.md

```diff
@@ -0,0 +1,104 @@
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace references exporting support.
+   */
+  xworkspaceReferencesProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace references request is sent from the client to the server to locate project-wide references to a symbol given its description / metadata.
+
+_Request_
+* method: 'workspace/xreferences'
+* params: `WorkspaceReferencesParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace References Request.
+ */
+interface WorkspaceReferencesParams {
+    /**
+     * Metadata about the symbol that is being searched for.
+     */
+    query: SymbolDescriptor;
+
+    /**
+     * An optional list of files to restrict the search to.
+     */
+    files?: string[];
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces, etc.
+ */
+interface ReferenceInformation {
+    /**
+     * The location in the workspace where the `symbol` has been referenced.
+     */
+    reference: Location;
+
+    /**
+     * Metadata information describing the symbol being referenced.
+     */
+    symbol: SymbolDescriptor;
+}
+```
+* error: code and message set in case an exception happens during the workspace references request.
+
+Where `SymbolDescriptor` is defined as follows:
+
+```typescript
+/**
+ * Represents information about a programming construct like a variable, class,
+ * interface, etc that has a reference to it. It is up to the language server
+ * to define the schema of this object.
+ *
+ * SymbolDescriptor usually uniquely identifies a symbol, but it is
+ * not guaranteed to do so.
+ */
+interface SymbolDescriptor {
+    [attr: string]: any
+}
+```
+
+### Goto Definition Extension Request
+
+This method is the same as `textDocument/definition`, except that:
+
+1. The method returns metadata about the definition (the same metadata that `workspace/xreferences` searches for).
+2. The concrete location to the definition (`location` field) is optional. This is useful because the language server might not be able to resolve a goto definition request to a concrete location (e.g. due to lack of dependencies) but still may know _some_ information about it.
+
+_Request_
+* method: 'textDocument/xdefinition'
+* params: [`TextDocumentPositionParams`](#textdocumentpositionparams)
+
+_Response_:
+* result: `LocationInformation[]` defined as follows:
+```typescript
+interface LocationInformation {
```

Comment:
I don't think that would be good. The actual field describing the definition is called `symbol` -- so we'd have to name it `SymbolInformation` which.. sounds a bit like something, you know? ;)

Also we're saying this is a `Location` OR `Information` which IMO makes sense.